weave/trace/weave_client_send_file_cache.py

```diff
@@ -0,0 +1,171 @@
+"""
+In memory cache of files sent to the server to avoid sending the same file
+multiple times, using an LRU (Least Recently Used) eviction policy.
+"""
+
+import hashlib
+import threading
+from collections import OrderedDict
+from concurrent.futures import Future
+from typing import Generic, Optional, TypeVar
+
+from weave.trace_server.trace_server_interface import FileCreateReq, FileCreateRes
+
+# Define generic type variables
+K = TypeVar("K")  # Key type
+V = TypeVar("V")  # Value type
+
+
+class ThreadSafeLRUCache(Generic[K, V]):
+    """Thread-safe LRU cache implementation using OrderedDict."""
+
+    def __init__(self, max_size: int = 1000) -> None:
+        """Initialize an LRU cache with a maximum size.
+
+        Args:
+            max_size: Maximum number of items to store in the cache.
+                     If set to 0 or negative, cache size is unlimited.
+        """
+        self._max_size = max(0, max_size)  # Ensure max_size is non-negative
+        self._cache: OrderedDict[K, V] = OrderedDict()
```

Comment:
This is unnecessary.  Dicts are defined to be ordered in the python language spec as of py37